File: Algorithms/C++/Medium/148_Sort_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }

        ListNode* mid = get_middle(head);
        ListNode* right = mid->next;
        mid->next = NULL;

        ListNode* sorted_left = sortList(head);
        ListNode* sorted_right = sortList(right);

        return merge(sorted_left, sorted_right);
    }
private:
    ListNode* get_middle(ListNode* head) {
        if (head == NULL) {
            return NULL;
        }

        ListNode* fast = head->next;
        ListNode* slow = head;

        while (fast != NULL && fast->next != NULL) {
            fast = fast->next->next;
            slow = slow->next;
        }

        return slow;
    }

    ListNode* merge(ListNode* headA, ListNode* headB) {
        ListNode dummy(0);
        ListNode* node = &dummy;

        while (headA != NULL && headB != NULL) {
            if (headA->val < headB->val) {
                node->next = headA;
                headA = headA->next;
            } else {
                node->next = headB;
                headB = headB->next;
            }
            node = node->next;
        }

        if (headA != NULL) {
            node->next = headA;
        }
        if (headB != NULL) {
            node->next = headB;
        }

        return dummy.next;
    }
};

File: Algorithms/JavaScript/Medium/148_Sort_List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function(head) {
    if (head === null) {
        return null
    }

    return mergeSort(head)
};

var mergeSort = function(head) {
    if (head === null || head.next === null) {
        return head
    }

    var mid = getMiddle(head)
    var left = head
    var right = mid.next
    mid.next = null

    var sortedLeft = mergeSort(left)
    var sortedRight = mergeSort(right)

    return merge(sortedLeft, sortedRight)
}

var getMiddle = function(head) {
    if (head === null) {
        return null
    }

    var slow = head
    var fast = head.next

    while (fast !== null && fast.next !== null) {
        slow = slow.next
        fast = fast.next.next
    }

    return slow
}

var merge = function(headA, headB) {
    var dummy = new ListNode(0)

    var node = dummy
    while (headA !== null && headB !== null) {
        if (headA.val < headB.val) {
            node.next = headA
            headA = headA.next
        } else {
            node.next = headB
            headB = headB.next
        }
        node = node.next
    }

    if (headA !== null) {
        node.next = headA
    }
    if (headB !== null) {
        node.next = headB
    }

    return dummy.next
}
